chain/ethereum/src/trigger.rs

```diff
@@ -229,11 +229,75 @@ impl ToAscPtr for MappingTrigger {
     }
 }
 
+#[derive(Clone, Debug)]
+pub enum LogRef {
+    FullLog(Arc<Log>, Option<Arc<TransactionReceipt>>),
+    LogPosition(usize, Arc<TransactionReceipt>),
+}
+
+impl LogRef {
+    pub fn log_index(&self) -> U256 {
+        match self {
+            LogRef::FullLog(log, _) => log.log_index.unwrap_or(U256::zero()),
+            LogRef::LogPosition(index, receipt) => receipt
+                .logs
+                .get(*index)
+                .unwrap()
+                .log_index
+                .unwrap_or(U256::zero()),
+        }
+    }
+
+    pub fn transaction_index(&self) -> Option<U64> {
+        match self {
+            LogRef::FullLog(log, _) => log.transaction_index,
+            LogRef::LogPosition(_, receipt) => Some(receipt.transaction_index),
+        }
+    }
+
+    fn transaction_hash(&self) -> Option<H256> {
+        match self {
+            LogRef::FullLog(log, _) => log.transaction_hash,
+            LogRef::LogPosition(_, receipt) => Some(receipt.transaction_hash),
+        }
+    }
+
+    pub fn receipt(&self) -> Option<&Arc<TransactionReceipt>> {
+        match self {
+            LogRef::FullLog(_, receipt) => receipt.as_ref(),
+            LogRef::LogPosition(_, receipt) => Some(receipt),
+        }
+    }
+
+    pub fn log(&self) -> Arc<Log> {
+        match self {
+            LogRef::FullLog(log, _) => log.clone(),
+            LogRef::LogPosition(index, receipt) => {
+                Arc::new(receipt.logs.get(*index).unwrap().clone())
+            }
+        }
+    }
+}
+
+impl PartialEq for LogRef {
```

Comment:
Where is this impl being relied on exactly?

Reply:
https://github.com/graphprotocol/graph-node/blob/b1fa9cecd4f6951788bb32dc3e151db0eb66acba/chain/ethereum/src/trigger.rs#L303

Its relied in `impl PartialEq for EthereumTrigger`

Reply:
I find it weird that this compares the `receipt` field given that it is optional, seems wrong really. Lets not have this impl, and inline the logic into `EthereumTrigger` as it was before. That code is critical for correct ordering of the triggers so I find that less indirection makes it easier to read. Also lets stop comparing the `receipt`, because it's optional, and instead compare only on `transaction_hash` and `log_index`.

Reply:
@leoyvens clarifying i understand this right, should we remove the previously existing comparison of the optional receipt right? Or should we just inline the logic and keep the original comparison of the optional receipt?

Reply:
done! 

I'll run it through the integration cluster